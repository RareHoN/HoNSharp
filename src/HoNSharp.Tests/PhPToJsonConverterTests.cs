using ApprovalTests;
using ApprovalTests.Reporters;
using Newtonsoft.Json;
using HoNSharp.Converters;
using HoNSharp.Models;

namespace HoNSharp.Tests
{
    [UseReporter(typeof(DiffReporter))]
    public class PhPToJsonConverterTests
    {
        private readonly PhPToJsonConverter _converter;

        public PhPToJsonConverterTests()
        {
            _converter = new PhPToJsonConverter();
        }

        [Fact]
        public void ConvertToJson_GetAccountAllHeroStats_ShouldCorrectlyConvert()
        {
            // Arrange
            var responseFileContent = ReadFile("get_account_all_hero_stats_response.txt");

            // Act
            var result = _converter.ConvertToJson(responseFileContent);

            // Assert
            Approvals.Verify(FormatJson(result));

            _ = JsonConvert.DeserializeObject<AllHeroStatsResponse>(result);
        }

        [Fact]
        public void ConvertToJson_GetMatchStats_ShouldCorrectlyConvert()
        {
            // Arrange
            var responseFileContent = ReadFile("get_match_stats_response.txt");

            // Act
            var result = _converter.ConvertToJson(responseFileContent);

            // Assert
            Approvals.Verify(FormatJson(result));

            _ = JsonConvert.DeserializeObject<MatchStatsResponse>(result);
        }

        [Fact]
        public void ConvertToJson_GetMatchStats_WithAnotherError_ShouldCorrectlyConvert()
        {
            // Arrange
            var responseFileContent = ReadFile("get_match_stats_another_error_response.txt");

            // Act
            var result = _converter.ConvertToJson(responseFileContent);

            // Assert
            Approvals.Verify(FormatJson(result));

            _ = JsonConvert.DeserializeObject<MatchStatsResponse>(result);
        }

        [Fact]
        public void ConvertToJson_GetMatchStats_WithDateTimeOffsetError_ShouldCorrectlyConvert()
        {
            // Arrange
            var responseFileContent = ReadFile("get_match_stats_datetimeoffset_error_response.txt");

            // Act
            var result = _converter.ConvertToJson(responseFileContent);

            // Assert
            Approvals.Verify(FormatJson(result));

            _ = JsonConvert.DeserializeObject<MatchStatsResponse>(result);
        }

        [Fact]
        public void ConvertToJson_GetMatchStats_WithUnexpectedCharacterError_ShouldCorrectlyConvert()
        {
            // Arrange
            var responseFileContent = ReadFile("get_match_stats_unexpected_character_error_response.txt");

            // Act
            var result = _converter.ConvertToJson(responseFileContent);

            // Assert
            Approvals.Verify(FormatJson(result));

            _ = JsonConvert.DeserializeObject<MatchStatsResponse>(result);
        }

        [Fact]
        public void ConvertToJson_MatchHistoryOverviewMidwars_ShouldCorrectlyConvert()
        {
            // Arrange
            var responseFileContent = ReadFile("match_history_overview_midwars_response.txt");

            // Act
            var result = _converter.ConvertToJson(responseFileContent);

            // Assert
            Approvals.Verify(FormatJson(result));

            _ = JsonConvert.DeserializeObject<Dictionary<string, string>>(result);
        }

        [Fact]
        public void ConvertToJson_MatchHistoryOverviewNormal_ShouldCorrectlyConvert()
        {
            // Arrange
            var responseFileContent = ReadFile("match_history_overview_normal_response.txt");

            // Act
            var result = _converter.ConvertToJson(responseFileContent);

            // Assert
            Approvals.Verify(FormatJson(result));

            _ = JsonConvert.DeserializeObject<Dictionary<string, string>>(result);
        }

        [Fact]
        public void ConvertToJson_MatchHistoryOverviewEmpty_ShouldCorrectlyConvert()
        {
            // Arrange
            var responseFileContent = ReadFile("match_history_overview_empty_response.txt");

            // Act
            var result = _converter.ConvertToJson(responseFileContent);

            // Assert
            Approvals.Verify(FormatJson(result));

            _ = JsonConvert.DeserializeObject<Dictionary<string, string>>(result);
        }

        [Fact]
        public void ConvertToJson_ShowStatsCampaign_ShouldCorrectlyConvert()
        {
            // Arrange
            var responseFileContent = ReadFile("show_stats_campaign_response.txt");

            // Act
            var result = _converter.ConvertToJson(responseFileContent);

            // Assert
            Approvals.Verify(FormatJson(result));

            _ = JsonConvert.DeserializeObject<CampaignStatsResponse>(result);
        }

        [Fact]
        public void ConvertToJson_ShowStatsCampaign_WithLongConvertError_ShouldCorrectlyConvert()
        {
            // Arrange
            var responseFileContent = ReadFile("show_stats_campaign_long_convert_error_response.txt");

            // Act
            var result = _converter.ConvertToJson(responseFileContent);

            // Assert
            Approvals.Verify(FormatJson(result));

            _ = JsonConvert.DeserializeObject<CampaignStatsResponse>(result);
        }

        [Fact]
        public void ConvertToJson_ShowStatsCampaignError_ShouldCorrectlyConvert()
        {
            // Arrange
            var responseFileContent = ReadFile("show_stats_campaign_error_response.txt");

            // Act
            var result = _converter.ConvertToJson(responseFileContent);

            // Assert
            Approvals.Verify(FormatJson(result));

            _ = JsonConvert.DeserializeObject<CampaignStatsResponse>(result);
        }

        [Fact]
        public void ConvertToJson_ShowStatsCampaignMidwars_ShouldCorrectlyConvert()
        {
            // Arrange
            var responseFileContent = ReadFile("show_stats_campaign_midwars_response.txt");

            // Act
            var result = _converter.ConvertToJson(responseFileContent);

            // Assert
            Approvals.Verify(FormatJson(result));

            _ = JsonConvert.DeserializeObject<CampaignStatsResponse>(result);
        }

        [Fact]
        public void ConvertToJson_ShowStatsMastery_ShouldCorrectlyConvert()
        {
            // Arrange
            var responseFileContent = ReadFile("show_stats_mastery_response.txt");

            // Act
            var result = _converter.ConvertToJson(responseFileContent);

            // Assert
            Approvals.Verify(FormatJson(result));

            _ = JsonConvert.DeserializeObject<MasteryStatsResponse>(result);
        }

        [Fact]
        public void ConvertToJson_SimpleStats_ShouldCorrectlyConvert()
        {
            // Arrange
            var responseFileContent = ReadFile("simple_stats_response.txt");

            // Act
            var result = _converter.ConvertToJson(responseFileContent);

            // Assert
            Approvals.Verify(FormatJson(result));

            _ = JsonConvert.DeserializeObject<SimpleStatsResponse>(result);
        }

        [Fact]
        public void ConvertToJson_PlayerAwardSumm_ShouldCorrectlyConvert()
        {
            // Arrange
            var responseFileContent = ReadFile("get_player_award_summ_response.txt");

            // Act
            var result = _converter.ConvertToJson(responseFileContent);

            // Assert
            Approvals.Verify(FormatJson(result));

            _ = JsonConvert.DeserializeObject<AwardSummaryResponse>(result);
        }

        private string ReadFile(string responseFileName)
        {
            var responseFilePath = Path.Combine(Directory.GetCurrentDirectory(), "TestData", responseFileName);

            return File.ReadAllText(responseFilePath);
        }

        private string FormatJson(string json)
        {
            var deserialized = JsonConvert.DeserializeObject(json);
            return JsonConvert.SerializeObject(deserialized, Formatting.Indented);
        }
    }
}