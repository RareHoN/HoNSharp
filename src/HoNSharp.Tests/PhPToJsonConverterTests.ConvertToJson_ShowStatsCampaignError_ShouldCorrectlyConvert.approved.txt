{
  "super_id": 2086,
  "nickname": "Anubis2108",
  "standing": 3,
  "account_type": 4,
  "account_id": 2086,
  "level": 1,
  "level_exp": 0,
  "discos": 0,
  "possible_discos": 0,
  "games_played": 329,
  "num_bot_games_won": 0,
  "total_games_played": 329,
  "total_discos": 0,
  "acc_secs": 626919,
  "acc_games_played": 329,
  "acc_discos": 0,
  "rnk_secs": 0,
  "rnk_games_played": 0,
  "rnk_discos": 0,
  "cs_secs": 0,
  "cs_games_played": 0,
  "cs_discos": 0,
  "mid_games_played": 0,
  "mid_discos": 0,
  "rift_games_played": 0,
  "rift_discos": 0,
  "last_activity": "8/18/2022",
  "create_date": "8/18/2022",
  "season": 22,
  "cam_games_played": 329,
  "cam_wins": 146,
  "cam_losses": 183,
  "cam_concedes": 147,
  "cam_concedevotes": 119,
  "cam_buybacks": 14,
  "cam_discos": 0,
  "cam_kicked": 0,
  "cam_amm_solo_rating": {
    "MinValue": {
      "MinValue": null,
      "Epsilon": {
        "MinValue": null,
        "Epsilon": null,
        "MaxValue": {
          "MinValue": null,
          "Epsilon": null,
          "MaxValue": null,
          "PositiveInfinity": {
            "MinValue": null,
            "Epsilon": null,
            "MaxValue": null,
            "PositiveInfinity": null,
            "NegativeInfinity": {
              "MinValue": null,
              "Epsilon": null,
              "MaxValue": null,
              "PositiveInfinity": null,
              "NegativeInfinity": null,
              "NaN": {
                "MinValue": null,
                "Epsilon": null,
                "MaxValue": null,
                "PositiveInfinity": null,
                "NegativeInfinity": null,
                "NaN": null
              }
            },
            "NaN": null
          },
          "NegativeInfinity": null,
          "NaN": null
        },
        "PositiveInfinity": null,
        "NegativeInfinity": null,
        "NaN": null
      },
      "MaxValue": null,
      "PositiveInfinity": null,
      "NegativeInfinity": null,
      "NaN": null
    },
    "Epsilon": null,
    "MaxValue": null,
    "PositiveInfinity": null,
    "NegativeInfinity": null,
    "NaN": null
  },
  "cam_amm_solo_count": 0,
  "cam_amm_solo_conf": 0,
  "cam_amm_solo_prov": 0,
  "cam_amm_solo_pset": 0,
  "cam_amm_team_rating": null,
  "cam_amm_team_count": 329,
  "cam_amm_team_conf": 0,
  "cam_amm_team_prov": 0,
  "cam_amm_team_pset": 0,
  "cam_herokills": 1387,
  "cam_herodmg": 3095366,
  "cam_heroexp": 1634371,
  "cam_herokillsgold": 914001,
  "cam_heroassists": 2287,
  "cam_deaths": 1654,
  "cam_goldlost2death": 453628,
  "cam_secs_dead": 82789,
  "cam_teamcreepkills": 28404,
  "cam_teamcreepdmg": 10807874,
  "cam_teamcreepexp": 2119674,
  "cam_teamcreepgold": 1221428,
  "cam_neutralcreepkills": 12925,
  "cam_neutralcreepdmg": 9233513,
  "cam_neutralcreepexp": 796290,
  "cam_neutralcreepgold": 447367,
  "cam_bdmg": 232906,
  "cam_bdmgexp": 0,
  "cam_razed": 205,
  "cam_bgold": 413100,
  "cam_denies": 1884,
  "cam_exp_denied": 81803,
  "cam_gold": 3284000,
  "cam_gold_spent": 3240245,
  "cam_exp": 4556970,
  "cam_actions": 1071748,
  "cam_secs": 626919,
  "cam_consumables": 4640,
  "cam_wards": 456,
  "cam_em_played": 0,
  "cam_level": 1,
  "cam_level_exp": 0,
  "cam_min_exp": 0,
  "cam_max_exp": 0,
  "cam_time_earning_exp": 622285,
  "cam_bloodlust": 25,
  "cam_doublekill": 131,
  "cam_triplekill": 12,
  "cam_quadkill": 2,
  "cam_annihilation": 0,
  "cam_ks3": 119,
  "cam_ks4": 73,
  "cam_ks5": 43,
  "cam_ks6": 31,
  "cam_ks7": 15,
  "cam_ks8": 9,
  "cam_ks9": 6,
  "cam_ks10": 3,
  "cam_ks15": 0,
  "cam_smackdown": 0,
  "cam_humiliation": 7,
  "cam_nemesis": 854,
  "cam_retribution": 20,
  "acc_pub_skill": 1167,
  "acc_wins": 146,
  "acc_losses": 183,
  "acc_concedes": 147,
  "acc_concedevotes": 119,
  "acc_buybacks": 14,
  "acc_kicked": 0,
  "acc_amm_solo_rating": null,
  "acc_amm_solo_count": 0,
  "acc_amm_solo_conf": 0,
  "acc_amm_solo_prov": 0,
  "acc_amm_solo_pset": 0,
  "acc_amm_team_rating": null,
  "acc_amm_team_count": 329,
  "acc_amm_team_conf": 0,
  "acc_amm_team_prov": 0,
  "acc_amm_team_pset": 0,
  "acc_herokills": 1387,
  "acc_herodmg": 3095366,
  "acc_heroexp": 1634371,
  "acc_herokillsgold": 914001,
  "acc_heroassists": 2287,
  "acc_deaths": 1654,
  "acc_goldlost2death": 453628,
  "acc_secs_dead": 82789,
  "acc_teamcreepkills": 28404,
  "acc_teamcreepdmg": 10807874,
  "acc_teamcreepexp": 2119674,
  "acc_teamcreepgold": 1221428,
  "acc_neutralcreepkills": 12925,
  "acc_neutralcreepdmg": 9233513,
  "acc_neutralcreepexp": 796290,
  "acc_neutralcreepgold": 447367,
  "acc_bdmg": 232906,
  "acc_bdmgexp": 0,
  "acc_razed": 205,
  "acc_bgold": 413100,
  "acc_denies": 1884,
  "acc_exp_denied": 81803,
  "acc_gold": 3284000,
  "acc_gold_spent": 3240245,
  "acc_exp": 4556970,
  "acc_actions": 1071748,
  "acc_consumables": 4640,
  "acc_wards": 456,
  "acc_em_played": 0,
  "acc_level": 1,
  "acc_level_exp": 0,
  "acc_min_exp": 0,
  "acc_max_exp": 0,
  "acc_time_earning_exp": 622285,
  "acc_bloodlust": 25,
  "acc_doublekill": 131,
  "acc_triplekill": 12,
  "acc_quadkill": 2,
  "acc_annihilation": 0,
  "acc_ks3": 119,
  "acc_ks4": 73,
  "acc_ks5": 43,
  "acc_ks6": 31,
  "acc_ks7": 15,
  "acc_ks8": 9,
  "acc_ks9": 6,
  "acc_ks10": 3,
  "acc_ks15": 0,
  "acc_smackdown": 0,
  "acc_humiliation": 7,
  "acc_nemesis": 854,
  "acc_retribution": 20,
  "cam_cs_games_played": 329,
  "cam_cs_wins": 146,
  "cam_cs_losses": 183,
  "cam_cs_concedes": 147,
  "cam_cs_concedevotes": 119,
  "cam_cs_buybacks": 14,
  "cam_cs_discos": 0,
  "cam_cs_kicked": 0,
  "cam_cs_amm_solo_rating": null,
  "cam_cs_amm_solo_count": 0,
  "cam_cs_amm_solo_conf": 0,
  "cam_cs_amm_solo_prov": 0,
  "cam_cs_amm_solo_pset": 0,
  "cam_cs_amm_team_rating": null,
  "cam_cs_amm_team_count": 329,
  "cam_cs_amm_team_conf": 0,
  "cam_cs_amm_team_prov": 0,
  "cam_cs_amm_team_pset": 0,
  "cam_cs_herokills": 1387,
  "cam_cs_herodmg": 3095366,
  "cam_cs_heroexp": 1634371,
  "cam_cs_herokillsgold": 914001,
  "cam_cs_heroassists": 2287,
  "cam_cs_deaths": 1654,
  "cam_cs_goldlost2death": 453628,
  "cam_cs_secam_cs_dead": 82789,
  "cam_cs_teamcreepkills": 28404,
  "cam_cs_teamcreepdmg": 10807874,
  "cam_cs_teamcreepexp": 2119674,
  "cam_cs_teamcreepgold": 1221428,
  "cam_cs_neutralcreepkills": 12925,
  "cam_cs_neutralcreepdmg": 9233513,
  "cam_cs_neutralcreepexp": 796290,
  "cam_cs_neutralcreepgold": 447367,
  "cam_cs_bdmg": 232906,
  "cam_cs_bdmgexp": 0,
  "cam_cs_razed": 205,
  "cam_cs_bgold": 413100,
  "cam_cs_denies": 1884,
  "cam_cs_exp_denied": 81803,
  "cam_cs_gold": 3284000,
  "cam_cs_gold_spent": 3240245,
  "cam_cs_exp": 4556970,
  "cam_cs_actions": 1071748,
  "cam_cs_secs": 626919,
  "cam_cs_consumables": 4640,
  "cam_cs_wards": 456,
  "cam_cs_em_played": 0,
  "cam_cs_level": 1,
  "cam_cs_level_exp": 0,
  "cam_cs_min_exp": 0,
  "cam_cs_max_exp": 0,
  "cam_cs_time_earning_exp": 622285,
  "cam_cs_bloodlust": 25,
  "cam_cs_doublekill": 131,
  "cam_cs_triplekill": 12,
  "cam_cs_quadkill": 2,
  "cam_cs_annihilation": 0,
  "cam_cs_ks3": 119,
  "cam_cs_ks4": 73,
  "cam_cs_ks5": 43,
  "cam_cs_ks6": 31,
  "cam_cs_ks7": 15,
  "cam_cs_ks8": 9,
  "cam_cs_ks9": 6,
  "cam_cs_ks10": 3,
  "cam_cs_ks15": 0,
  "cam_cs_smackdown": 0,
  "cam_cs_humiliation": 7,
  "cam_cs_nemesis": 854,
  "cam_cs_retribution": 20,
  "name": null,
  "rank": "None",
  "favHero1": "devourer",
  "favHero1Time": 10.334346771240234,
  "favHero1_2": "Hero_Devourer",
  "favHero2": "valkyrie",
  "favHero2Time": 6.07902717590332,
  "favHero2_2": "Hero_Valkyrie",
  "favHero3": "doctor_repulsor",
  "favHero3Time": 4.55927038192749,
  "favHero3_2": "Hero_DoctorRepulsor",
  "favHero4": "magmar",
  "favHero4Time": 3.03951358795166,
  "favHero4_2": "Hero_Magmar",
  "favHero5": "rocky",
  "favHero5Time": 3.03951358795166,
  "favHero5_2": "Hero_Rocky",
  "dice_tokens": 1,
  "season_level": 0,
  "slot_id": 5,
  "my_upgrades": {
    "0": "m.allmodes.pass",
    "1": "h.AllHeroes.Hero",
    "2": "m.Super-Taunt"
  },
  "selected_upgrades": {
    "0": "ai.3 Seconds",
    "1": "av.Badass",
    "2": "cs.Rhapsody (CN)",
    "3": "t.Chiprel_Taunt"
  },
  "game_tokens": 0,
  "my_upgrades_info": {},
  "creep_level": 0,
  "timestamp": 1668286078,
  "matchIds": "000003989 000004214 000007322 000011048 000011525 000013428 000013937 000014236 000014405 000017687 000017835 000018083 000018541 000019753 000020344 000020465 000020584 000021364 000023150 000023290",
  "matchDates": "08/19/202208/19/202208/20/202208/21/202208/21/202208/22/202208/22/202208/22/202208/22/202208/23/202208/23/202208/23/202208/23/202208/24/202208/24/202208/24/202208/24/202208/24/202208/25/202208/25/2022",
  "k_d_a": "4/5/6",
  "avgGameLength": 1905.5289306640625,
  "avgXP_min": 439.3777770996094,
  "avgDenies": 5.726443767547607,
  "avgCreepKills": 86.33434295654297,
  "avgNeutralKills": 39.28571319580078,
  "avgActions_min": 102.57286834716797,
  "avgWardsUsed": 1.3860182762145996,
  "quest_stats": {
    "error": {
      "quest_status": 0,
      "leaderboard_status": 0
    }
  },
  "highest_level_current": 0,
  "current_level": 0,
  "level_percent": 93,
  "season_id": 22,
  "prev_seasons_cam_games_played": 0,
  "prev_seasons_cam_discos": 0,
  "latest_season_cam_games_played": 329,
  "latest_season_cam_discos": 0,
  "curr_season_cam_games_played": 329,
  "curr_season_cam_discos": 0,
  "prev_seasons_cam_cs_games_played": 0,
  "prev_seasons_cam_cs_discos": 0,
  "latest_season_cam_cs_games_played": 329,
  "latest_season_cam_cs_discos": 0,
  "curr_season_cam_cs_games_played": 329,
  "curr_season_cam_cs_discos": 0,
  "con_reward": {
    "old_lvl": 0,
    "curr_lvl": 0,
    "next_lvl": 0,
    "require_rank": 0,
    "need_more_play": 0,
    "percentage_before": "0.00",
    "percentage": "0.00"
  },
  "vested_threshold": 5,
  "0": true
}